Move stat_linux to cpu_linux and add cpufreq stats

rtreffer opened the pull request:
This PR is the result of the discussion in #538 

It moves the stat_linux collector to cpu_linux (as that's the key metric exported by the other cpu_<os> exporters) and add cpufreq data there.

discordianfish commented on collector/cpu_linux.go:
```diff
@@ -83,8 +114,64 @@ func NewStatCollector() (Collector, error) {
+	}
+	if err2 != nil {
+		return err2
+	}
```
Maybe just use the same err variable?
```
err := c.UpdateStat(ch)
err = c.UpdateCpufreq(ch)
if err ... 
```
Not sure if this is ideal but imo slightly better.

rtreffer replied:
I always want to report an error if either breaks, but I always want to try both.... Doesn't reusing the variable shadow the first error?

rtreffer replied:
But yeah, this is ugly, let's see if I can simplify it

discordianfish replied:
Oh right, you would need to check the error after each assignment. Still could use one though. Or something like:
```
var err error
for e := range []error{c.UpdateStat(ch)} {
  if e != nil {
    err = e
  }
}
```

But arguably this isn't better.. We might want to think about this case in general "partial failures in collectors".. Either way, it's also fine to merge this as is.

grobie replied:
I don't think using a for loop helps readability. So far we've mostly followed an all or nothing approach I think and aborted after the first error.

I'm fine with the approach above. You could use more descriptive names than a generic counter like `errStat`, but given the few lines involved it doesn't matter too much.

discordianfish commented:
Looks good in general but it's a breaking change as it is. I think we should print a deprecation warning and enable the cpu collector when 'stat' collector is enabled.

juliusv commented on collector/cpu_linux.go:
```diff
@@ -83,8 +114,64 @@ func NewStatCollector() (Collector, error) {
+}
+
+// Expose cpu frequency statistics.
+func (c *cpuCollector) UpdateCpufreq(ch chan<- prometheus.Metric) error {
```
UpdateCpuFreq -> updateCPUFreq (https://github.com/golang/go/wiki/CodeReviewComments#initialisms and should be unexported)

juliusv commented on collector/cpu_linux.go:
```diff
@@ -83,8 +114,64 @@ func NewStatCollector() (Collector, error) {
+
 // Expose kernel and system statistics.
-func (c *statCollector) Update(ch chan<- prometheus.Metric) error {
+func (c *cpuCollector) UpdateStat(ch chan<- prometheus.Metric) error {
```
UpdateStat -> updateStat

juliusv commented on collector/cpu_linux.go:
```diff
@@ -28,28 +29,58 @@ const (
-func NewStatCollector() (Collector, error) {
-	return &statCollector{
+// NewCpuCollector returns a new Collector exposing kernel/system statistics.
+func NewCpuCollector() (Collector, error) {
```
NewCpuCollector -> NewCPUCollector

juliusv commented on README.md:
```diff
@@ -37,7 +37,6 @@ meminfo | Exposes memory statistics. | Darwin, Dragonfly, FreeBSD, Linux
 netdev | Exposes network interface statistics such as bytes transferred. | Darwin, Dragonfly, FreeBSD, Linux, OpenBSD
 netstat | Exposes network statistics from `/proc/net/netstat`. This is the same information as `netstat -s`. | Linux
 sockstat | Exposes various statistics from `/proc/net/sockstat`. | Linux
-stat | Exposes various statistics from `/proc/stat`. This includes CPU usage, boot time, forks and interrupts. | Linux
```
Bit confusing that to get the boot time and some other CPU-unrelated stats, you now have to enable the CPU collector? But I also don't have a great answer to this question, maybe that's just the way it is...

discordianfish replied:
Hrm good point.. I think best would be to keep non-cpu stuff in the stat collector, if we can do that without much complication/code duplication.
@rtreffer What do you think?

This also reminds me of @derekmarcotte's refactoring suggestions which would be helpful here. But let's keep things simple for now.

grobie replied:
I vote to only move the CPU part and leave the rest in stat collector for now. It'd be confusing to control boot time etc. with the CPU collector.

grobie commented on collector/cpu_linux.go:
```diff
@@ -28,28 +29,58 @@ const (
 			[]string{"cpu", "mode"}, nil,
 		),
+		cpuFreq: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "cpu", "frequency_hertz"),
```
Let's use a constant for the repeated `cpu` string.

grobie commented on collector/cpu_linux.go:
```diff
@@ -83,8 +114,64 @@ func NewStatCollector() (Collector, error) {
+	return nil
+}
+
+// Expose cpu frequency statistics.
```
Go's docstring convention says it should be a `comment is a complete sentence that begins with the name of the element it describes.` https://blog.golang.org/godoc-documenting-go-code.

grobie commented on collector/cpu_linux.go:
```diff
@@ -83,8 +114,64 @@ func NewStatCollector() (Collector, error) {
+	return nil
+}
+
 // Expose kernel and system statistics.
```
See above regarding how to write function docstrings.

grobie commented on collector/cpu_linux.go:
```diff
@@ -28,28 +29,58 @@ const (
+func NewCpuCollector() (Collector, error) {
+	return &cpuCollector{
 		cpu: prometheus.NewDesc(
 			prometheus.BuildFQName(Namespace, "", "cpu"),
```
Hm, a metric which is the prefix of other metrics is a first and could be confusing. Changing that is outside of the scope of this PR, but we should finally decide on a timeline to rename all these metrics.

SuperQ commented:
Please rebase against master to pick up the fixes for the test pipeline and fix merge conflicts.

rtreffer commented:
I've update the PR based on the comments.

1. parsing of /proc/stat is now done with the procfs package
2. cpu collector (enabled by default) exports cpu related data on linux
3. stat collector exports non-cpu related /proc/stat data
4. Updated go docs and fixed some minor comments

Can I get a few :eyes: again?

SuperQ commented:
Looks good so far, but what about moving the cpufreq stuff to procfs?

grobie commented:
I wouldn't make this a requirement. At some point we also need to keep in
mind the time of our contributors. We've held back this pr for a long time
already.

On Fri, Jun 9, 2017, 16:48 Ben Kochie <notifications@github.com> wrote:

> Looks good so far, but what about moving the cpufreq stuff to procfs?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/node_exporter/pull/548#issuecomment-307409342>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaKkD_mV56fGMQLZzpwU3MPvUSQDnks5sCVszgaJpZM4M2EiF>
> .
>

SuperQ commented:
@grobie Totally agree, we can try and find some time to do a bunch of `procfs` refactoring later.

mdlayher commented:
Happy to help with that project ^.

discordianfish commented on collector/cpu_linux.go:
```diff
@@ -0,0 +1,160 @@
+
+		if value, err := readUintFromFile(filepath.Join(cpu, "cpufreq/scaling_cur_freq")); err != nil {
+			return err
+		} else {
```
Sorry for that nitpick, but can you drop the } else { block? That's more idiomatic. Like:

```
if value, err := readUintFromFile(filepath.Join(cpu, "cpufreq/scaling_cur_freq")); err != nil {
￼    return err
}
￼ch <- prometheus.MustNewConstMetric(c.cpuFreq, prometheus.GaugeValue, float64(value), cpuname)
```

rtreffer replied:
Fixed. Requires `value` to be defined at an outer scope.

rtreffer commented:
@discordianfish anything else I should change or can you change your review status?

discordianfish commented:
Looks good, great! Thanks!

grobie commented:
This change will need a `[CHANGE]` line in the changelog and probably an extra description to remind people specifying collectors manually to enable the cpu collector on linux now.
